### ISSUE TITLE
Potential Memory Leak in expand_as

### ISSUE THREAD

**Issue by zihangdai** [pytorch/pytorch#1963]
I think I encountered a memory leak situation potentially due to `expand_as`. I create a [gist](https://gist.github.com/zihangdai/5d4b6fa96ce1366f5bc7268fee0d4e9f) to reproduce my observations.

Basically, for three different (meaningless) models, once the `expand_as` function is used, the memory usage will keep increasing. 

To run the code, `python mem_leak_expand.py --run_expand --model_type [0,1,2]` where the `--run_expand` flag controls whether `expand_as` is used. Please run it long enough. 

I'm on Ubuntu 14.04 with python3.6. I tried two versions of pytorch, 0.1.12_2 and 0.1.12+da0fad8. 

**Side Note**: Even without using `expand_as`, the memory usauge can also increase for a while, but it will finally stablize at a level. What's more, when `Conv2d` (i.e. `--model_type [1,2]`) is used, it can take very long before the memory usage stablizes. I'm not sure whether this is a problem or not.

**Comment by zihangdai** [comment]
I think I'm sure there is a memory leak due to `expand_as`, and I found a weird way to do a temporary fix. The simplest example is as follows:
```python
from __future__ import print_function, division
import gc
import argparse
import resource

import torch
from torch.autograd import Variable

def main(args):
    tgt_var = Variable(torch.rand(32, 512))

    for i in range(args.num_steps):
        src_var = Variable(torch.rand(1, 512))
        if args.fix:
            expanded = src_var.expand(*list(tgt_var.size()))
        else:
            expanded = src_var.expand_as(tgt_var)

        if args.backward:
            ret.mean().backward()

        gc.collect()
        if i > 0 and i % 500 == 0:
            max_mem_used = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
            print("#{} {:.4f} MB".format(i, max_mem_used / 1024))

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='PyTorch expand_as memory leak')
    parser.add_argument('--num_steps', type=int, default=1000000)
    parser.add_argument('--backward', action='store_true', default=False)
    parser.add_argument('--fix', action='store_true', default=False)
    args = parser.parse_args()

    main(args)
```

**Comment by zihangdai** [comment]
After some more research, I think we have a bigger problem here. My current conclusion is that: whenever we use `*tensor.size()` as an unpacked list of arguments to a function, there will be a memory leak. 

An example is the `pad_packed_sequence` function in `torch/nn/utils/rnn.py`, where [line 107](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/rnn.py#L107) of the code is `output = var_data.data.new(len(batch_sizes), max_batch_size, *var_data.size()[1:]).zero_()` and uses `*var_data.size()`. 

Again, a minimal example for reproduction is as follows:

```python
from __future__ import print_function, division
import gc
import argparse
import resource

import torch
import torch.nn as nn
from torch.autograd import Variable

import torch.nn.utils.rnn as rnn_utils

def main(args):
    max_len = 8
    hid_dim = 16
    batch_s = 4
    model = nn.GRU(hid_dim, hid_dim)
    for i in range(args.num_steps):
        
        embedding = Variable(torch.rand(max_len, batch_s, hid_dim))
        lengths = range(max_len, max_len-batch_s, -1)
        
        packed_emb = rnn_utils.pack_padded_sequence(embedding, lengths)
        packed_out, hidden_final = model(packed_emb)
        outputs, srclens = rnn_utils.pad_packed_sequence(packed_out)

        if args.backward:
            outputs.mean().backward()

        gc.collect()
        if i > 0 and i % 500 == 0:
            max_mem_used = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
            print("#{} {:.4f} MB".format(i, max_mem_used / 1024))

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='PyTorch *tensor.size() memory leak')
    parser.add_argument('--num_steps', type=int, default=1000000)
    parser.add_argument('--backward', action='store_true', default=False)
    args = parser.parse_args()

    main(args)
```

As in the `expand_as` case, once I change the code to `output = var_data.data.new(len(batch_sizes), max_batch_size, *list(var_data.size())[1:]).zero_()`, memory leak is gone. 

So, in short, the problem is on the usage of `tensor.size()`.

**Comment by donglixp** [comment]
The same issue. +1

**Comment by colesbury** [comment]
Thanks @zihangdai for tracking this down and for the repro script!

**Comment by zihangdai** [comment]
@colesbury Thanks for the quick fix! BTW, will the fix be included in the release version (0.1.12_?)?

**Comment by soumith** [comment]
@zihangdai it'll be released in `0.2.0`

**Comment by zihangdai** [comment]
@soumith I see. Thanks again.
